app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java

```diff
@@ -50,51 +60,112 @@ public Bitmap convert(final InputStream inputStream) throws IOException {
 
     @Override
     public void draw(final String pUrl, final ImageView pView) {
-        draw(pUrl, pView, pView.getWidth(), pView.getHeight());
+        draw(pUrl, pView, null);
     }
 
     @Override
-    public void draw(final String pUrl, final ImageView pView, final int pWidth, final int pHeight) {
+    public void draw(final String pUrl, final ImageView pView, final DisplayOptions pOptions) {
+        if (pUrl == null) {
+            setPlaceholder(pView, pOptions.getPlaceholderId());
+
+            return;
+        }
+
+        if (pUrl.equals(pView.getTag())) {
+            return;
+        }
+
+        if (pOptions == null || pOptions.getWidth() == 0 || pOptions.getHeight() == 0) {
+            final DisplayOptions options;
+
+            if (pOptions == null) {
+                options = new DisplayOptions(0);
+            } else {
+                options = pOptions;
+            }
+
+            pView.post(new Runnable() {
+
+                @Override
+                public void run() {
+                    options.setWidth(pView.getWidth());
+                    options.setHeight(pView.getHeight());
+                    execute(pUrl, pView, options);
+                }
+            });
+        } else {
+            execute(pUrl, pView, pOptions);
+        }
+    }
+
+    private void execute(final String pUrl, final ImageView pView, @NonNull final DisplayOptions pOptions) {
+        final WeakReference<ImageView> imageViewReference = new WeakReference<>(pView);
+        setPlaceholder(pView, pOptions.getPlaceholderId());
         pView.setTag(pUrl);
-        final Handler handler = new Handler();
-        new Thread(new Runnable() {
+
+        mExecutorService.execute(new Runnable() {
 
             @Override
             public void run() {
-                mSizeParamsCallback = initSizeParamsCallback(pWidth, pHeight);
+                if (imageViewReference.get() == null || !pUrl.equals(imageViewReference.get().getTag())) {
+                    return;
+                }
+
+                mSizeParamsCallback = initSizeParamsCallback(pOptions.getWidth(), pOptions.getHeight());
 
                 IHttpClient.Impl.get().doRequest(pUrl, new IHttpClient.IOnResult<Bitmap>() {
 
                             @Override
-                            public void onSuccess(final Bitmap pBitmap) {
-                                final Bitmap bitmap = resizeBitmap(pBitmap, pWidth, pHeight);
-                                
-                                handler.post(new Runnable() {
+                            public void onSuccess(Bitmap pBitmap) {
+                                final Bitmap bitmap = centerCropBitmap(pBitmap, pOptions.getWidth(), pOptions.getHeight());
```

Comment:
what if I need a circle bitmap? or real-size? 
Again, lets return to this when really needed